### DISCUSSION TITLE
Is this how laravel will look like in v15?

### DISCUSSION THREAD

**Discussion by ghost** [laravel/framework#53917] [Q&A] [answered]
(This is a pamphlet, so please don't be offended)

I see more and more features making their way into the core, especially the ones that use php attributes, so I ask you:

 Is this how laravel will look like in version 15?

![image](https://github.com/user-attachments/assets/b05211e0-575e-4d6d-9366-46667e223b39)

**Comment by Sophist-UK** [comment, marked as answer]
When the alternatives are only:

* features in core Laravel; or
* 3rd party packages for very commonly used functionality + difficulties in finding and choosing from competing packages amongst the mass of small community packages

then perhaps features in core Laravel makes sense.

And remember that Taylor is extremely resistant to feature growth for _exactly_ this reason. Imagine what it would be like if he wasn't so resistant??!!

But perhaps it might be possible to split out the core but less used functionality into packages and have artisan commands to make it easy to add them in (assuming that we don't want to expect the user to run a `composer` command instead).

**Reply by ghost** [reply]
May your words become law in laravel community.

**Reply by Sophist-UK** [reply]
I wish I understood what that meant. 😄

**Reply by ghost** [reply]
I meant that laravel should not become bloated with features.
I wish it would evolve towards slimmer core and packages that are not interconnected.
**I still use lumen because of this reason.**
APIs don't need all that laravel offers in its core. 
Laravel is behaving like all projects are monoliths and should continue to exist as monoliths.

Laravel **octane is cutting away features from php** (static) and from DI (singletons). This is... no comment.
And the community is arguing that because of how fast is octane, laravel can handle countless reflections for attributes etc etc.
I use even the workers with **--once** to avoid that in memory booted app.

**Comment by ahinkle** [comment]
There isn't a reason to use Lumen when you have Octane.

Laravel isn’t "bloated with features" — it’s flexible and built to handle a variety of use cases. You can easily use only the parts you need for APIs without the extras. No idea where you got that Octane is removing features. I've seen them instead trying to improve performance with better persistent memory. 

It’s not pushing monoliths, it’s about giving developers options.

**Reply by ghost** [reply]
To avoid bugs generated by static and singletons,  I refuse to use octane. Being unable to use all that php has to offer (because it is introducing unwanted behavior - which I understand by the way) to me means octane is cutting those features in exchange for speed.

**Comment by ghost** [comment]
I dream of a laravel free of politics that will not release new versions each year.

I dream to have a bare bone inspired by Laravel/lumen 10 latest version at most (not 11) that should be maintained until the maintenance is impossible without breaking changes.

New features should come only via packages and once every 5 years lets say (medium lifetime of a project) should a new version be released with breaking changes and their new features but only if needed.

Benefits.
People will not have to update each year to a new version but just update BACKWARDS COMPATIBLE bug fixes and improvements in v 10.
When starting a new project from scratch, people could choose from multiple stable maintained versions depending on their needs.

No more milking clients for their money for useless updates and of course no more big profits for devs and their contractors.

Last thought.
People tend to think that if they need a feature, it MUST go in the core, like they own laravel. This is a bad bad idea.
Their needs must stay in their project not in laravel/lumen. At most they can create a package for them.

If you (laravel community) don't wake up, you will end up just like in the above picture and younger devs will stop and say:
Wait! Why should I use this framework when I can create one for my own needs that will be faster, smaller, more efficient, with better codding standards and so on. Or maybe there is such an alternative already.

**Reply by henzeb** [reply]
That is why Taylor is denying so many PR's... You really need to chill out. Also, Laravel is free. Never had to pay for any upgrade in my life. And if you follow the paradigms closely and write good tests, upgrading doesn't take much of your time either.

**Reply by ghost** [reply]
Clients that use laravel in their businesses need to pay developers to update laravel. Just like they have to pay to update wordpress for example. That is what I meant. If you write unit tests, you ~ double the development time => you double the cost for the clients.

**Reply by shaedrich** [reply]
Now we're getting quite dreamy—and I like it! 😍 👍🏻 

I wouldn't agree with everything to 100 %, but I agree with the overall sentiment that I'd like to see a general change of direction.

My dreams:
<ul>
<li>I don't have a problem with PRs getting closed for <strong>good reasons</strong>. And I don't even say, there are none in most cases. There probably are. But one of my biggest pain points is Taylor's poor to non-existent (exercising of) communication skills. Many say, he's a nice guy and that might very well be. But it is said that "the function of a thing is what it does" and therefore, that is what I see what he does: Copy-pasting generic response templates, sometimes even not quite fitting into PRs as closing reason if even one is given with everyone left to guess and speculate about his specific reason. Besides, being frustrating, this has a couple more severe problems:

<ul>
<li>Subsequent PRs get opened with what is assumed to be better fitting for Taylor, yet, nobody knows for certain. This results in more workload for Taylor, which I assume, he already has quite plenty of</li>
<li>Perfectly valid, but unfortunately badly timed changes are not resubmitted at a later time, because the author assumes, these changes are not wanted at all</li>
<li>Devs might not even become contributors, because they think, nothing they would contribute would even make it to the framework, since they have no way of telling how their changes might be received</li>
</ul>

and especially said workload can be challenging. Maybe, this is too much for one man to handle. But the solution to it should not be to rush things by either approving too many PRs (he doesn't do that) or by rejecting too many (he might be more prone to that but it might just look like it, because: <em>communication</em>). What would help, would be

<ul>
<li>Prioritizing and clearly communicating the milestones, maybe even the overall focus for the next major version</li>
<li>if he can't do this alone, can't other core team members support him in approving/rejecting PRs?</li>
<li>if he is just not a good communicator, maybe, Laravel needs someone better at it to support each other. This would not be a bad thing, because we all have different things we are better or not so good at.</li>
</ul>

</li>
<li>I agree, not everything has to part of the core, many frameworks have parts of this or all three:

<ul>
<li><strong>Core:</strong> A lightweight (maybe even bare metal) base package (like Lumen) with everything you need to use Laravel</li>
<li><strong>Standard library:</strong> Laravel's first-party packages</li>
<li><strong>Third-party extensions:</strong> Other composer packages</li>
</ul>

Laravel has all that <em>in general</em>, but more could be done to encourage people to use the best approach for their needs:

<ul>
<li>
<strong>Clear instructions:</strong> When a PR is rejected, because it is better suited as a package

<ul>
<li>try to add a short line as to why this is. This especially helps unexperienced developers and might lead to better subsequent contributions to the framework</li>
<li>add a link to the docs, so people can jump right there without having to first go to the Laravel website (It's not much, but some might loose interest or feel left alone without)</li>
<li>encourage them to share the package and add tips on how to spread the word of the new package and potential upgrading it to a first-party package (<em>if the possibility is seen at this point</em>)</li>
</ul>

</li>
<li><strong>Clear and easy extension points:</strong> Extending a framework can be done in various way, some are easier to do with PHP alone, some harder, some can be done via composer but don't feel very neat.</li>
<li>
<strong>Package managing:</strong> Just like composer, pear, npm, apt and many others, Laravel needs a way to organize packages to be easily discoverable and installable for developers

<ul>
<li><strong>"App store":</strong> An official site to browse first- and third-party packages with good categorization and metadata to make it easy to decide what you need</li>
<li><strong>Install command:</strong> Laravel currently usually installs packages with composer and commands like <code>php artisan {package}:install</code>. Would be easier to have a command like <code>php artisan install:{package}</code> with autocomplete, filtering and such. This could even be enhanced by the prompts packages and perhaps an interactive TUI</li>
</ul>

</li>
</ul>

</li>
<li>What is also likely to confuse new contributors is how things are meant to be done around here, which can kinda result in opinionated edit wars and, despite linting, a somewhat inconsistent coding style</li>
</ul>

**Reply by henzeb** [reply]
> If you write unit tests, you ~ double the development time => you double the cost for the clients.

Sure, but if you don't they have to pay developers to fix their own bugs and you need a tester ...

> Clients that use laravel in their businesses need to pay developers to update laravel.

Doesn't really matter as you will always have to ask developers to update / maintain the versions anyway. Unless you have clients that can do this themselves. Keep in mind that even while minor and patch updates generally do not introduce breaking changes, they can cause problems that can corrupt data... I don't want to do things that break my stuff I can't fix and the developers aren't available right away...

**Reply by Sophist-UK** [reply]
> If you write unit tests, you ~ double the development time => you double the cost for the clients.

All this comment proves is just how little you understand about the purpose of tests.

Believe it or not, the vast majority of decisions made in this world are done on a financial basis - and the decisions to write tests are absolutely made on this basis. Tests make software much much more maintainable, and that means that whilst the development costs are a bit higher, the maintenance costs are MUCH lower.

Just think about how much effort goes into testing a new IT system - and the number of test cycles that need to be done to fix bugs and retest the system - and how much that costs. Automated unit tests iron out a whole bunch of bugs that would otherwise make it through to system testing and cost a fortune to fix.

Also think about how many software issues later in production can be traced to poor quality changes to an existing application introducing new bugs, and the consequential costs to the business that are incurred, and how much effort goes into tracking down this bug and fixing it. In more cases than not, a comprehensive pre-written suite of unit tests would have prevented that bug making into into production.

**Reply by ghost** [reply]
@Sophist-UK If you refer to integration tests then we are on the same page. If you talk only about UNIT tests, then I've been there, done that, thank you, never again I will waste my time writing UNIT tests. They are just a marketing scheme for milking money form the clients. There are dozens of articles online describing their uselessness better than I can describe it. The amount of help they bring, in real world is close to 0, especially in an business where the business logic is changing often. Your unit tested code is useless if it is delivered after the client goes bankrupt.

**Reply by henzeb** [reply]
That really depends on the situation. You don't have to have 100% coverage, you need to test at least the critical parts of the business. The parts where your clients lose money when it fails. The login, the payment flow, for example. You don't need to test the for all possible things that can fail, just the most critical ones. And then you can add tests when bug fixing... 

And surely, if you have many clients, I am sure they share a lot of the same code. If that code is tested well, you are half way there...

Tests aren't useless. They giving devs more confidence in releasing it. They are a form of proof that it works. But yes, again, when written well.

**Reply by ghost** [reply]
@henzeb 

> That really depends on the situation.

Your words are gold. The parts that you just mentioned must be covered with integration tests not with unit tests. Testing is very important, but it must be done in the most efficient way and in a bulletproof way. Unit tests can't offer you that because they work in a mocked environment, funny enough, written by the same dev who wrote the code. 

I had the luck of learning (when I entered in this field) from a team of professionals working for a huge client on rewriting a huge and old app, (with unit tests). The positive side of it was that they showed me how to write code that does not need unit testing (to see the typos, bugs etc). It is something laravel codding style does not have and I will not go into details about it.

In an academical world, unit tests are a must. In a war zone (the real world is an economical war at least), unit tests are a rock on your foot.

**Reply by henzeb** [reply]
Integration tests are also kind of mocked, as they assume a start and an end situation. I think they are more work to set up.

Unit tests are definitely not a rock on your foot if you use them wisely. They can help you to understand better where something is broken. An integration test does that not so much...

And we don't mock all unit tests . They are given an input and an expected output.

**Reply by ghost** [reply]
@henzeb if you prefer to use unit tests to check for bugs in your code, it is your decision. They can help in that situation but, from there to saying that they ensure the project works as intended after a laravel update for example... that is far from being realistic. This picture sums it up very well: **_plausible deniability_**
![image](https://github.com/user-attachments/assets/56353caa-34e4-4fd3-b7c5-38967274cda8)

**Reply by Sophist-UK** [reply]
That has got to be one of the most biased and bigoted memes I have ever seen. 😉 

(I have revised this post to take into account the points in the Youtube video linked below.)

1. We need to clarify what we mean by "Unit" tests - what is the Unit? The Youtube video says a module is a Unit not a class or a method - and this might align with Domain Driven Design. But somehow this falls apart with Microservices where a Domain is a self contained service and what might be System tests in a monolithic development are integration tests with Microservices, and Modular tests in a monolithic system are effectively System tests with microservices. So perhaps we need to agree on definitions first, and I would propose Integration (end-to-end), Module and Unit tests, where Unit tests are class/method level.

2. I definitely agree that focussing unit tests on critical components is higher priority than trying to create blanket coverage across all methods of all objects.

3. Given a choice between automated system tests and automated unit tests, I would absolutely go for System tests, but they do need to be comprehensive.

4. Tests are only as good as the things that they test:
   * basic functionality, rare cases only or every code path - IMO readable code is more important than tests here - basic functionality will soon become obvious when wrong and should be picked up by system tests, but rare cases might be something worth unit testing.
   * programmatic exception cases - to ensure that the method doesn't fail with programmatically incorrect input - IMO better handled by strict typing
   * user input exception cases - you ask for an integer, they enter an alpha - better handled by only allowing numbers to be input in the input field mask, but if an API then should be tested - but test in validation and NOT in downstream methods which should rely on strict typing
   * edge cases - if the user is supposed to enter a decimal number between 1 and 5, then 0.99999 fails, 1 passes, 5 passes, 5.00000001 fails

5. TDD (obviously) relies on tests being written first. Definitely applicable to testing big chunks of code (or system tests), of dubious net benefit when applied to low-level methods.

6. Unit tests are apparently an excellent use case for AI generation - but Garbage-in = Garbage-out as tests generated from possibly erroneous source methods will pass - so the tests need to be QA'd against the specs too. So if you want low level unit tests for regression purposes, then using AI to generate them close to final acceptance might not be a bad idea.

**Reply by shaedrich** [reply]
https://www.youtube.com/watch?v=EZ05e7EMOLM

**Reply by ghost** [reply]
@shaedrich 
Exactly this (your youtube link) is what I meant by:
> There are dozens of articles online describing their uselessness better than I can describe it.

I stopped watching when I saw the hexagonal (architecture) figure.

**Reply by Sophist-UK** [reply]
I am watching this Youtube video now, and this guy is not wrong.
